node/cn/state_accessor.go

```diff
@@ -162,11 +156,8 @@ func (cn *CN) stateAtBlock(block *types.Block, reexec uint64, base *state.StateD
 		if cn.config.DisableUnsafeDebug && time.Since(start) > cn.config.StateRegenerationTimeLimit {
 			return nil, nil, fmt.Errorf("this request has queried old states too long since it exceeds the state regeneration time limit(%s)", cn.config.StateRegenerationTimeLimit.String())
 		}
-		// Preload StakingInfo from the current block and state. Needed for next block's engine.Finalize() post-Kaia.
-		preloadedStakingBlockNums = append(preloadedStakingBlockNums, current.NumberU64())
-		if err := reward.PreloadStakingInfoWithState(current.Header(), statedb); err != nil {
-			return nil, nil, fmt.Errorf("preloading staking info from block %d failed: %v", current.NumberU64(), err)
-		}
+		// Make StakingModule remember the current block state. Needed for next block's engine.Finalize() post-Kaia.
+		cn.stakingModule.AddSideState(sideStateRef, statedb)
```

Comment:
If any error occurs below, `sideStateRef` can't be released. It seems previous implementation also had same issue.

Reply:
Let me work on it.

Reply:
suggestion: use named return for error, and a defer func checks if it is not nil, in which case free the ref.

Reply:
Fixed in 22ea3e7c.

By the way, Deference() is also missing in some error cases. But because geth has the same issue I won't fix that here. https://github.com/ethereum/go-ethereum/blob/v1.14.11/eth/state_accessor.go

Reply:
cc @ian0371 @yoomee1313

Reply:
The state root is always dereferenced regardless of error. If no error, the caller dereferences it. Other cases is done here: https://github.com/kaiachain/kaia/blob/1fe135d9e12ab4ec8057aa5e71b6423353eb7834/node/cn/state_accessor.go#L188-L190

So our case is different. The releaser won't be run by the caller when there's error, so the callee must call it in defer function.

Reply:
Oh, I see. the `root` is quite short-lived.

The release() function doing the FreePreloadRef was an attempt to replace `reward.PreloadStakingInfo` with `stateAtBlock` so we have only one implementation. But I later learned that merging them was non-trivial.

I'll go back to the original way of calling `defer FreePreloadRef()` inside stateAtBlock.